// @snapshot-labs/checkpoint/dist/src/schemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkpointConfigSchema = exports.contractTemplateSchema = exports.contractSourceConfigSchema = exports.contractEventConfigSchema = void 0;
const zod_1 = require("zod");
exports.contractEventConfigSchema = zod_1.z.object({
    name: zod_1.z.string(),
    fn: zod_1.z.string()
});
exports.contractSourceConfigSchema = zod_1.z.object({
    contract: zod_1.z.string(),
    abi: zod_1.z.string().optional(),
    start: zod_1.z.number().gte(0),
    deploy_fn: zod_1.z.string().optional(),
    events: zod_1.z.array(exports.contractEventConfigSchema)
});
exports.contractTemplateSchema = zod_1.z.object({
    abi: zod_1.z.string().optional(),
    events: zod_1.z.array(exports.contractEventConfigSchema)
});
exports.checkpointConfigSchema = zod_1.z.object({
    network_node_url: zod_1.z.string().url(),
    optimistic_indexing: zod_1.z.boolean().optional(),
    decimal_types: zod_1.z
        .record(zod_1.z.object({
        p: zod_1.z.number(),
        d: zod_1.z.number()
    }))
        .optional(),
    start: zod_1.z.number().gte(0).optional(),
    tx_fn: zod_1.z.string().optional(),
    global_events: zod_1.z.array(exports.contractEventConfigSchema).optional(),
    sources: zod_1.z.array(exports.contractSourceConfigSchema).optional(),
    templates: zod_1.z.record(exports.contractTemplateSchema).optional()
});
