// @snapshot-labs/checkpoint/dist/src/utils/checkpoint.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfigChecksum = exports.getContractsFromConfig = void 0;
const object_hash_1 = __importDefault(require("object-hash"));
const getContractsFromConfig = (config) => {
    return (config.sources || []).map(source => source.contract);
};
exports.getContractsFromConfig = getContractsFromConfig;
const getHashableProperties = (config) => ({
    contract: 'contract' in config ? config.contract : undefined,
    start: 'start' in config ? config.start : undefined,
    events: (config.events || []).map(event => event.name)
});
const getConfigChecksum = (config) => {
    const { tx_fn, global_events, sources, templates } = config;
    return (0, object_hash_1.default)({
        tx_fn,
        global_events,
        sources: (sources || []).map(source => getHashableProperties(source)),
        templates: Object.fromEntries(Object.entries(templates || {}).map(([key, value]) => [key, getHashableProperties(value)]))
    }, {
        unorderedArrays: true
    });
};
exports.getConfigChecksum = getConfigChecksum;
