// @snapshot-labs/checkpoint/dist/src/utils/graphql.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNonNullType = exports.generateQueryForEntity = exports.multiEntityQueryName = exports.singleEntityQueryName = exports.extendSchema = void 0;
const graphql_1 = require("graphql");
const json_to_graphql_query_1 = require("json-to-graphql-query");
const pluralize_1 = __importDefault(require("pluralize"));
const extendSchema = (schema) => {
    return `directive @derivedFrom(field: String!) on FIELD_DEFINITION
${schema}`;
};
exports.extendSchema = extendSchema;
/**
 * Returns name of query for fetching single entity record
 *
 */
const singleEntityQueryName = (entity) => entity.name.toLowerCase();
exports.singleEntityQueryName = singleEntityQueryName;
/**
 * Returns name of query for fetching multiple entity records
 *
 */
const multiEntityQueryName = (entity) => {
    if (entity.name === '_Metadata')
        return '_metadatas';
    return (0, pluralize_1.default)(entity.name.toLowerCase());
};
exports.multiEntityQueryName = multiEntityQueryName;
/**
 * Generate sample query string based on entity object fields.
 *
 */
const generateQueryForEntity = (entity) => {
    // function to recursively build fields map
    const getObjectFields = (object, queryFields = {}) => {
        const objectFields = object.getFields();
        Object.keys(objectFields).forEach(fieldName => {
            const rawFieldType = objectFields[fieldName].type;
            const fieldType = rawFieldType instanceof graphql_1.GraphQLNonNull ? rawFieldType.ofType : rawFieldType;
            if ((0, graphql_1.isLeafType)(fieldType)) {
                queryFields[fieldName] = true;
            }
            else if ((0, graphql_1.isListType)(fieldType)) {
                if (fieldType.ofType instanceof graphql_1.GraphQLScalarType) {
                    queryFields[fieldName] = true;
                }
            }
            else {
                const childObjectFields = {};
                getObjectFields(fieldType, childObjectFields);
                queryFields[fieldName] = childObjectFields;
            }
        });
        return queryFields;
    };
    return (0, json_to_graphql_query_1.jsonToGraphQLQuery)({
        query: {
            [(0, exports.multiEntityQueryName)(entity)]: {
                __args: { first: 10 },
                ...getObjectFields(entity)
            }
        }
    }, { pretty: true });
};
exports.generateQueryForEntity = generateQueryForEntity;
const getNonNullType = (type) => {
    if (type instanceof graphql_1.GraphQLNonNull) {
        return type.ofType;
    }
    return type;
};
exports.getNonNullType = getNonNullType;
