// @snapshot-labs/sx/dist/clients/starknet/starknet-sig/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StarkNetSig = void 0;
const cross_fetch_1 = __importDefault(require("cross-fetch"));
const types_1 = require("./types");
class StarkNetSig {
    address;
    constructor(address) {
        this.address = address;
    }
    async sign(signer, address, message, types, primaryType) {
        const data = { domain: types_1.domain, types, primaryType, message };
        const sig = await signer.signMessage(data);
        return { address, sig, data };
    }
    async send(envelop) {
        const init = {
            method: 'POST',
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                jsonrpc: '2.0',
                method: 'send',
                params: { envelop },
                id: null
            })
        };
        const res = await (0, cross_fetch_1.default)(this.address, init);
        const json = await res.json();
        return json.result;
    }
    async propose(web3, address, message) {
        return await this.sign(web3, address, message, types_1.proposeTypes, 'Propose');
    }
    async vote(web3, address, message) {
        return await this.sign(web3, address, message, types_1.voteTypes, 'Vote');
    }
}
exports.StarkNetSig = StarkNetSig;
