// @snapshot-labs/sx/dist/utils/strategies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getExtraProposeCalls = exports.getStrategiesParams = exports.getStrategies = void 0;
const starknet_1 = require("../strategies/starknet");
const encoding_1 = require("../utils/encoding");
async function getStrategies(data, config) {
    const addresses = await Promise.all(data.strategies.map(id => config.starkProvider.getStorageAt(data.space, (0, encoding_1.getStorageVarAddress)('Voting_voting_strategies_store', id.toString(16)))));
    return data.strategies.map((v, i) => ({
        index: v,
        address: addresses[i]
    }));
}
exports.getStrategies = getStrategies;
async function getStrategiesParams(call, strategies, address, data, config) {
    return Promise.all(strategies.map(strategyData => {
        const strategy = (0, starknet_1.getStrategy)(strategyData.address, config.networkConfig);
        if (!strategy)
            throw new Error('Invalid strategy');
        return strategy.getParams(call, strategyData.address, strategyData.index, {
            address,
            sig: '',
            data: {
                message: data
            }
        }, config);
    }));
}
exports.getStrategiesParams = getStrategiesParams;
async function getExtraProposeCalls(strategies, address, data, config) {
    const extraCalls = await Promise.all(strategies.map(strategyData => {
        const strategy = (0, starknet_1.getStrategy)(strategyData.address, config.networkConfig);
        if (!strategy)
            throw new Error('Invalid strategy');
        return strategy.getExtraProposeCalls(address, strategyData.index, {
            address,
            sig: '',
            data: {
                message: data
            }
        }, config);
    }));
    return extraCalls.flat();
}
exports.getExtraProposeCalls = getExtraProposeCalls;
