// @snapshot-labs/sx/dist/authenticators/starknet/ethSig.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const encoding_1 = require("../../utils/encoding");
const split_uint256_1 = require("../../utils/split-uint256");
function createEthSigAuthenticator() {
    return {
        type: 'ethSig',
        createCall(envelope, selector, calldata) {
            const { sig, data } = envelope;
            const { space, authenticator, salt } = data.message;
            const { r, s, v } = (0, encoding_1.getRSVFromSig)(sig);
            const rawSalt = split_uint256_1.SplitUint256.fromHex(`0x${salt.toString(16)}`);
            return {
                contractAddress: authenticator,
                entrypoint: 'authenticate',
                calldata: [
                    r.low,
                    r.high,
                    s.low,
                    s.high,
                    v,
                    rawSalt.low,
                    rawSalt.high,
                    space,
                    selector,
                    calldata.length,
                    ...calldata
                ]
            };
        }
    };
}
exports.default = createEthSigAuthenticator;
